infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServerCache.java

```diff
@@ -1,52 +1,68 @@
  * Contains transaction conflict-resolution data structures.
  * <p>
- * The cache map maps conflict keys (stream id + key) to versions (long), illustrated below:
+ * The conflictKeyMap maps conflict keys (stream id + key) to versions (long), illustrated below:
  * Conflict Key | ck1 | ck2 | ck3 | ck4
  * Version | v1 | v1 | v2 | v3
- * Consider the case where we need to insert a new conflict key (ck), but the cache is full,
- * we need to evict the oldest conflict keys. In the example above, we can't just evict ck1,
- * we need to evict all keys that map to v1, so we need to evict ck1 and ck2,
- * this eviction policy is FIFO on the version number. The simple FIFO approach in Caffein etc doesn't work here,
- * as it may evict ck1, but not ck2. Notice that we also can't evict ck3 before the keys for v1,
- * that's because it will create holes in the resolution window and can lead to incorrect resolutions.
+ * Consider the case where we need to insert a new conflict key (ck), but the cache is full.
+ * We need to evict the oldest conflict keys. In the example above, we can't just evict ck1,
+ * we need to evict all keys that map to v1, so we need to evict ck1 and ck2. This eviction
+ * policy is FIFO on the version number. Notice that we also can't evict ck3 before the keys
+ * for v1, that's because it will create holes in the resolution window and can lead to
+ * incorrect resolutions.
  * <p>
- * We use priority queue as a sliding window on the versions, where a version can map to multiple keys,
+ * We use TreeMap as a sliding window on the versions, where a version can map to multiple keys,
  * so we also need to maintain the beginning of the window which is the maxConflictWildcard variable.
  * <p>
- * SequencerServerCache achieves consistency by using single threaded cache. It's done by following code:
- * `.executor(Runnable::run)`
+ * SequencerServerCache achieves consistency by using single threaded cache.
  */
 @NotThreadSafe
 @Slf4j
 public class SequencerServerCache implements AutoCloseable {
+
     /**
-     * TX conflict-resolution information:
-     * a cache of recent conflict keys and their latest global-log position.
+     * A mapping between conflict keys and their latest global-log position.
      */
+    @VisibleForTesting
+    @Getter(AccessLevel.PUBLIC)
+    private final Map<ConflictTxStream, Long> conflictKeyMap;
 
-    // As the sequencer cache is used by a single thread, it is safe to use hashmap.
-    private final HashMap<ConflictTxStream, Long> conflictKeys;
-    private final PriorityQueue<ConflictTxStream> cacheEntries; //sorted according to address
+    /**
+     * A mapping between global-log positions and conflict keys. This is maintained to
+     * facilitate the eviction of all conflict keys associated with a particular transaction
+     * commit timestamp.
+     */
+    @VisibleForTesting
+    @Getter(AccessLevel.PUBLIC)
+    private final SortedMap<Long, Set<ConflictTxStream>> versionMap;
+
```

Comment:
I don't think this is the best data structure to use especially when the entries are being added to the collection in sorted order. For example, `versionMap.put` is called by `cache.put(conflictTxStreamList, newTail - 1);`. Since newTail - 1 is strictly increasing, you end up adding sorted input to the `SortedMap` which is not necessary. 

Based on that invariant you can use a linked list to keep the sliding window versions. I think that would simply the implementation and make more efficient.

Reply:
Thanks for pointing this out. After more thought, realized that both data structures can be combined together by using a LinkedHashMap, and gives the best of both worlds.